### PR TITLE
[ZFE-30] Custom directive v-loading

### PR DESCRIPTION
**Pull request by WawrzenskiPatryk**

### CODE REVIEW

File: assets/css/directives.css [hunk 1]
```diff
@@ -0,0 +1,9 @@
+.v-loading {
+  @apply relative opacity-60 pointer-events-none select-none overflow-hidden min-h-[30px] min-w-[30px];
+}
+
+.v-loading::after {
+  @apply content-[''] absolute z-20 w-4 h-4 border-2 rounded-full animate-[spin_600ms_linear_infinite] border-black border-t-transparent;
+  top: calc(50% - 8px);
+  left: calc(50% - 8px);
```
Comment (WawrzenskiPatryk) [inline assets/css/directives.css:8]: Tried to add everything in tailwind, but this part is something I couldn't find a solution for

File: components/ProductsList.vue [hunk 2]
```diff
@@ -12,7 +12,10 @@
       </NuxtLink>
     </div>
 
-    <ul class="flex gap-10">
+    <ul
+      v-loading="isLoading"
+      class="flex gap-10"
+    >
```
Comment (WawrzenskiPatryk) [inline components/ProductsList.vue:18]: So this file is an example usage, to be implemented everywhere else

File: plugins/directives.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    mounted(el, { value }) {
+      setLoadingClass(el, value);
+    },
+    updated(el, { value }) {
+      setLoadingClass(el, value);
+    },
+    getSSRProps({ value }: LoadingBinding) {
+      return {
+        class: value || value === undefined ? LOADING_CLASS : '',
+      };
```
Comment (WawrzenskiPatryk) [inline plugins/directives.ts:29]: This part is what makes our directive to work also on server side :rocket:

File: types/directives.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import { ObjectDirective, DirectiveBinding } from 'nuxt/dist/app/compat/capi';
+
+export type LoadingValue = boolean | undefined;
+
+export type LoadingBinding = DirectiveBinding<LoadingValue>;
+
+export type LoadingDirective = ObjectDirective<HTMLElement, LoadingValue>;
```
Comment (WawrzenskiPatryk) [inline types/directives.ts:7]: Please tell me if what I put here is a good approach as this is my first serious attempt with TS

File: assets/css/directives.css (hunk 1 above)
Comment (WawrzenskiPatryk) [reply assets/css/directives.css]: Actually did it 👌

File: assets/css/directives.css [hunk 5]
```diff
@@ -0,0 +1,7 @@
+.v-loading {
+  @apply relative opacity-60 pointer-events-none select-none overflow-hidden min-h-[30px] min-w-[30px];
+}
+
+.v-loading::after {
+  @apply content-[''] absolute top-1/2 left-1/2 -mt-2 -ml-2 z-20 w-4 h-4 border-2 rounded-full animate-[spin_600ms_linear_infinite] border-black border-t-transparent;
```
Comment (tomekskuta-monterail) [inline assets/css/directives.css:6]: It's cool that you found a way to replace css rules with tailwind classes 🙌

File: plugins/directives.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import '@/assets/css/directives.css';
+import {
+  LoadingValue,
+  LoadingBinding,
+  LoadingDirective,
+} from '@/types/directives';
+
+export default defineNuxtPlugin((nuxt) => {
+  const LOADING_CLASS = 'v-loading';
```
Comment (tomekskuta-monterail) [inline plugins/directives.ts:9]: 💪

File: types/directives.ts (hunk 4 above)
Comment (tomekskuta-monterail) [reply types/directives.ts:7]: looks good 👌

**Review by tomekskuta-monterail — APPROVED** [review]

**Review by elmir-ismayilzada — APPROVED** [review]
Looks good 👍

**Review by zakrzaq — APPROVED** [review]
Looks really good! 🥇

**Pull request merged**